src/test_ambit.cc

```diff
@@ -0,0 +1,1672 @@
+
+    D4("ijkl") += batched("ijkl", A4("ijmn") * B2("km") * B2("ln"));
+
+    for (size_t i = 0; i < ni; ++i) {
+        for (size_t j = 0; j < nj; ++j) {
+            for (size_t k = 0; k < nk; ++k) {
+                for (size_t l = 0; l < nl; ++l) {
+                    for (size_t m = 0; m < nm; ++m) {
+                        for (size_t n = 0; n < nn; ++n) {
+                            d4[i][j][k][l] += a4[i][j][m][n] * b2[k][m] * b2[l][n];
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return difference(D4, d4).second;
+}
+
+double test_batched_with_factor() {
+    size_t no = 2;
+
+    ::std::vector<size_t> dimsA = {no, no, no, no};
+    ::std::vector<size_t> dimsB = {no, no, no, no};
+    ::std::vector<size_t> dimsC = {no, no, no, no};
+
+    Tensor A = build_and_fill("A", dimsA, a4);
+    Tensor B = build_and_fill("B", dimsB, b4);
+    Tensor C = build_and_fill("C", dimsC, c4);
+
+    C("ijrs") = batched("r", 0.5 * A("abrs") * B("ijab"));
+
+    for (size_t i = 0; i < no; ++i) {
+        for (size_t j = 0; j < no; ++j) {
+            for (size_t r = 0; r < no; ++r) {
+                for (size_t s = 0; s < no; ++s) {
+                    c4[i][j][r][s] = 0.0;
+                    for (size_t a = 0; a < no; ++a) {
+                        for (size_t b = 0; b < no; ++b) {
+                            c4[i][j][r][s] += 0.5 * a4[a][b][r][s] * b4[i][j][a][b];
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return difference(C, c4).second;
+}
+
+double test_batched_with_factor_permute() {
+    size_t no = 4;
+
+    ::std::vector<size_t> dimsA = {no, no, no, no};
+    ::std::vector<size_t> dimsB = {no, no, no, no};
+    ::std::vector<size_t> dimsC = {no, no, no, no};
+
+    Tensor A = build_and_fill("A", dimsA, a4);
+    Tensor B = build_and_fill("B", dimsB, b4);
+    Tensor C = build_and_fill("C", dimsC, c4);
+
+    C("ijrs") = batched("r", 0.5 * A("abrs") * B("ijba"));
+
+    for (size_t i = 0; i < no; ++i) {
+        for (size_t j = 0; j < no; ++j) {
+            for (size_t r = 0; r < no; ++r) {
+                for (size_t s = 0; s < no; ++s) {
+                    c4[i][j][r][s] = 0.0;
+                    for (size_t a = 0; a < no; ++a) {
+                        for (size_t b = 0; b < no; ++b) {
+                            c4[i][j][r][s] += 0.5 * a4[a][b][r][s] * b4[i][j][b][a];
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return difference(C, c4).second;
+}
+
+bool test_ambit() {
+    srand(time(nullptr));
+
+    auto test_functions = {
+        //            Expectation,  test function,  User friendly description
+        ::std::make_tuple(kPass, test_wrapper,
+                          "8 permutations of C(\"ij\") += A(\"ik\") * B(\"kj\")"),
+        ::std::make_tuple(kPass, test_C_equal_2_A, "C(\"ij\") = 2.0 * A(\"ij\")"),
```

Comment:
This looks a bit funny, why is the "::" prefix needed?

Reply:
Oh, please look at the current Travis failed check as I removed "::". I do not fully understand but this is what I had to do in https://github.com/evangelistalab/forte/pull/74/commits/3328564f2da0480b861d8a6ff2a6f60b1c2a047c